File: interview/dsa-drills/ruby/test/dsa_drills_test.rb
# frozen_string_literal: true

require "dsa_drills"

def assert_equal(expected, actual)
  return if expected == actual

  raise "Expected #{expected.inspect}, got #{actual.inspect}"
end

assert_equal [0, 1], DsaDrills.two_sum([2, 7, 11, 15], 9)
assert_equal [1, 2], DsaDrills.two_sum([3, 2, 4], 6)
assert_equal [], DsaDrills.two_sum([1, 2, 3], 10)

assert_equal 3, DsaDrills.length_of_longest_substring("abcabcbb")
assert_equal 1, DsaDrills.length_of_longest_substring("bbbbb")
assert_equal 3, DsaDrills.length_of_longest_substring("pwwkew")

assert_equal [24, 12, 8, 6], DsaDrills.product_except_self([1, 2, 3, 4])
assert_equal [0, 0, 9, 0, 0], DsaDrills.product_except_self([-1, 1, 0, -3, 3])

assert_equal 4, DsaDrills.search_rotated([4, 5, 6, 7, 0, 1, 2], 0)
assert_equal(-1, DsaDrills.search_rotated([4, 5, 6, 7, 0, 1, 2], 3))
assert_equal(-1, DsaDrills.search_rotated([], 3))

grid = [
  %w[1 1 0 0 0],
  %w[1 1 0 0 0],
  %w[0 0 1 0 0],
  %w[0 0 0 1 1]
]
assert_equal 3, DsaDrills.num_islands(grid)

assert_equal [1, 2], DsaDrills.top_k_frequent([1, 1, 1, 2, 2, 3], 2)
assert_equal [1], DsaDrills.top_k_frequent([1], 1)

cache = DsaDrills::LRUCache.new(2)
cache.put(1, 1)
cache.put(2, 2)
assert_equal 1, cache.get(1)
cache.put(3, 3)
assert_equal(-1, cache.get(2))
cache.put(4, 4)
assert_equal(-1, cache.get(1))
assert_equal 3, cache.get(3)
assert_equal 4, cache.get(4)

assert_equal 3, DsaDrills.coin_change([1, 2, 5], 11)
assert_equal(-1, DsaDrills.coin_change([2], 3))
assert_equal 0, DsaDrills.coin_change([1], 0)

puts "Ruby DSA drills passed"

File: interview/dsa-drills/typescript/test/dsaDrills.test.ts
import test from "node:test";
import assert from "node:assert/strict";

import {
  LRUCache,
  coinChange,
  lengthOfLongestSubstring,
  numIslands,
  productExceptSelf,
  searchRotated,
  topKFrequent,
  twoSum
} from "../src/dsaDrills";

test("twoSum", () => {
  assert.deepEqual(twoSum([2, 7, 11, 15], 9), [0, 1]);
  assert.deepEqual(twoSum([3, 2, 4], 6), [1, 2]);
  assert.deepEqual(twoSum([1, 2, 3], 10), []);
});

test("lengthOfLongestSubstring", () => {
  assert.equal(lengthOfLongestSubstring("abcabcbb"), 3);
  assert.equal(lengthOfLongestSubstring("bbbbb"), 1);
  assert.equal(lengthOfLongestSubstring("pwwkew"), 3);
});

test("productExceptSelf", () => {
  assert.deepEqual(productExceptSelf([1, 2, 3, 4]), [24, 12, 8, 6]);
  assert.deepEqual(productExceptSelf([-1, 1, 0, -3, 3]), [0, 0, 9, 0, 0]);
});

test("searchRotated", () => {
  assert.equal(searchRotated([4, 5, 6, 7, 0, 1, 2], 0), 4);
  assert.equal(searchRotated([4, 5, 6, 7, 0, 1, 2], 3), -1);
  assert.equal(searchRotated([], 3), -1);
});

test("numIslands", () => {
  const grid = [
    ["1", "1", "0", "0", "0"],
    ["1", "1", "0", "0", "0"],
    ["0", "0", "1", "0", "0"],
    ["0", "0", "0", "1", "1"]
  ];

  assert.equal(numIslands(grid), 3);
});

test("topKFrequent", () => {
  assert.deepEqual(topKFrequent([1, 1, 1, 2, 2, 3], 2), [1, 2]);
  assert.deepEqual(topKFrequent([1], 1), [1]);
});

test("LRUCache", () => {
  const cache = new LRUCache(2);

  cache.put(1, 1);
  cache.put(2, 2);
  assert.equal(cache.get(1), 1);

  cache.put(3, 3);
  assert.equal(cache.get(2), -1);

  cache.put(4, 4);
  assert.equal(cache.get(1), -1);
  assert.equal(cache.get(3), 3);
  assert.equal(cache.get(4), 4);
});

test("coinChange", () => {
  assert.equal(coinChange([1, 2, 5], 11), 3);
  assert.equal(coinChange([2], 3), -1);
  assert.equal(coinChange([1], 0), 0);
});
